modules/porous_flow/src/materials/PorousFlowBrine.C

```diff
@@ -0,0 +1,206 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/**********************************************2******************/
+
+#include "PorousFlowBrine.h"
+#include "Conversion.h"
+
+template<>
+InputParameters validParams<PorousFlowBrine>()
+{
+  InputParameters params = validParams<PorousFlowWater>();
+  params.addRequiredParam<Real>("xnacl", "The salt mass fraction in the brine (kg/kg)");
+  params.addClassDescription("This Material calculates fluid properties for brine");
+  return params;
+}
+
+PorousFlowBrine::PorousFlowBrine(const InputParameters & parameters) :
+    PorousFlowWater(parameters),
+
+    _Mnacl(58.443e-3),
+    _xnacl(getParam<Real>("xnacl"))
+{
+}
+
+void
+PorousFlowBrine::initQpStatefulProperties()
+{
+  _density_nodal[_qp] = density(_porepressure_nodal[_qp][_phase_num], _temperature_nodal[_qp][_phase_num], _xnacl);
+}
+
+void
+PorousFlowBrine::computeQpProperties()
+{
+  /// Density and derivatives wrt pressure and temperature at the nodes
+  _density_nodal[_qp] = density(_porepressure_nodal[_qp][_phase_num], _temperature_nodal[_qp][_phase_num], _xnacl);
+  _ddensity_nodal_dp[_qp] = dDensity_dP(_porepressure_nodal[_qp][_phase_num], _temperature_nodal[_qp][_phase_num], _xnacl);
+  _ddensity_nodal_dt[_qp] = dDensity_dT(_porepressure_nodal[_qp][_phase_num], _temperature_nodal[_qp][_phase_num], _xnacl);
+
+  /// Density and derivatives wrt pressure and temperature at the qps
+  _density_qp[_qp] = density(_porepressure_qp[_qp][_phase_num], _temperature_qp[_qp][_phase_num], _xnacl);
+  _ddensity_qp_dp[_qp] = dDensity_dP(_porepressure_qp[_qp][_phase_num], _temperature_qp[_qp][_phase_num], _xnacl);
+  _ddensity_qp_dt[_qp] = dDensity_dT(_porepressure_qp[_qp][_phase_num], _temperature_qp[_qp][_phase_num], _xnacl);
+
+  /// Viscosity and derivative wrt temperature at the nodes
+  _viscosity_nodal[_qp] = viscosity(_temperature_nodal[_qp][_phase_num], _density_nodal[_qp], _xnacl);
+  _dviscosity_nodal_dt[_qp] = dViscosity_dT(_temperature_nodal[_qp][_phase_num], _density_nodal[_qp], _xnacl);
+}
+
+Real
+PorousFlowBrine::density(Real pressure, Real temperature, Real xnacl) const
+{
+  Real n1, n2, n11, n12, n1x1, n20, n21, n22, n23, n2x1, Tv;
+  Real water_density, density;
+
+  // The correlation requires the pressure in bar, not Pa.
+  Real pbar = pressure * 1.0e-5;
+  Real pbar2 = pbar * pbar;
+  Real pbar3 = pbar2 * pbar;
+
+  // The correlation requires mole fraction. First calculate the average molar mass
+  // from the mass fraction, then compute the mole fraction
+  Real Mbrine = 1.0 / (xnacl / _Mnacl + (1.0 - xnacl) / _Mh2o);
+  Real Xnacl = xnacl / _Mnacl * Mbrine;
+
+  n11 = -54.2958 - 45.7623 * std::exp(-9.44785e-4 * pbar);
```

Comment:
just a suggestion, but maybe a literature reference would be helpful if somebody ever wants to trace the origin of all these numbers.